### DISCUSSION TITLE
Error in Set-PnPUserProfileProperty - You are not authorized

### DISCUSSION THREAD

**Discussion by Geo-Ron** [pnp/powershell#293] [Q&A]
Hi all,

Version: 1.3.0 PNP.Powershell Manifest

I am trying to connect to Sharepoint using App (Certificate) Authentication.
This works, uploading files goes perfect.

Now I want to update the user Profile property and I receive an access denied.

Interactive logon works with my global admin account.

I ran the App registration as specified in the documentation with `Register-PnPAzureADApp -ApplicationName PnPRocks -Tenant mytenant.onmicrosoft.com -OutPath c:\mycertificates -DeviceLogin`
Additionally granting permissions to the app.
![image](https://user-images.githubusercontent.com/23144316/108196598-91d97200-7119-11eb-8778-c89f575862be.png)

Afterwards I ran `Register-PnPManagementShellAccess` and verified the permissions on this Enterprise App.

But still can not get it to work.

Maybe related to https://github.com/pnp/powershell/issues/277 but I am unsure if I have followed the correct steps.

Any help would be greatly appreciated...

**Comment by veronicageek** [comment]
Hi @Geo-Ron,
I suggest we first wait for the results of Erwin's investigation and see if that answers your question.

**Reply by Geo-Ron** [reply]
Good suggestion.

I was only wondering if my approach was correct. This remains unanswered for now.

**Comment by Geo-Ron** [comment]
Any update on this? #277 is about to expire due to inactivity...

**Comment by Markarend** [comment]
Here's an answer that promises to solve this issue; the only problem for me is that I still have the issue. But maybe it's a clue.

[Accessing SharePoint using an application context, also known as app-only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly) says "User Profile CSOM write operations do not work with Azure AD application - read operations work. Both read and write operations work through SharePoint App-Only principal".  Then [Granting access using SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs) gives instructions for setting that up. I was feeling pretty good when I found this, but then when I tested, still "Access denied. You do not have permission to perform this action or access this resource. Microsoft.SharePoint.Client.ServerUnauthorizedAccessException, ServerErrorCode : -2147024891, ServerErrorTypeName : System.UnauthorizedAccessException".

Another possible clue is that [Granting access using SharePoint App-Only](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/security-apponly-azureacs) shows -AppId and -AppSecret parameters for Connect-PnPOnline, but the current version (1.3.0?) doesn't provide those parameters, only -ClientId and -ClientSecret.  

Note, this all works for Get-PnPUserProfileProperty but not Set-PnPUserProfileProperty.  Arrgh!
